src/alcazar.rs

```diff
@@ -1,42 +1,79 @@
-use crate::http_request::HttpRequest;
-use std::net::{SocketAddr, TcpListener};
+use crate::{
+    http_request::{HttpError, HttpRequest, ParseError},
+    router::Router,
+};
+use std::{
+    io::{Error as IOError, Write},
+    net::{IpAddr, Ipv4Addr, SocketAddr, TcpListener},
+};
+use thiserror::Error;
 use tracing::info;
 
-#[derive(Default)]
-pub struct AlcazarBuilder {
-    url: Option<SocketAddr>,
+#[derive(Error, Debug)]
+pub enum AlcazarError {
+    #[error(transparent)]
+    IOError(#[from] IOError),
+    #[error(transparent)]
+    HttpError(#[from] HttpError),
+    #[error(transparent)]
+    ParseError(#[from] ParseError),
 }
 
-impl AlcazarBuilder {
-    pub fn set_addr(self, url: SocketAddr) -> Self {
-        Self { url: url.into() }
+pub struct AppBuilder {
+    addr: SocketAddr,
+    router: Router,
+}
+
+impl Default for AppBuilder {
+    fn default() -> Self {
+        Self {
+            addr: SocketAddr::new(IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)), 0),
+            router: Router::default(),
+        }
     }
+}
 
-    pub fn start(self) -> Alcazar {
-        let listener =
-            TcpListener::bind(self.url.unwrap_or_else(|| "0.0.0.0:0".parse().unwrap())).unwrap();
+impl AppBuilder {
+    pub fn set_addr(&mut self, addr: SocketAddr) -> &mut Self {
+        self.addr = addr;
+        self
+    }
 
-        let local_addr = listener.local_addr().unwrap();
+    pub fn set_router(&mut self, router: Router) -> &mut Self {
+        self.router = router;
+        self
+    }
+
+    pub fn start(&self) -> Result<App, AlcazarError> {
+        let listener = TcpListener::bind(self.addr)?;
+        let local_addr = listener.local_addr()?;
+        let router = self.router.clone();
 
         info!("listening to {}", local_addr);
-        std::thread::spawn(move || loop {
-            match listener.accept() {
-                Ok((stream, _addr)) => {
-                    HttpRequest::parse_stream(stream);
+        std::thread::spawn(move || -> Result<(), AlcazarError> {
+            loop {
+                match listener.accept() {
+                    Ok((mut stream, _addr)) => {
+                        let http_request = HttpRequest::parse_stream(&stream)?;
+                        let handler =
+                            router.get_handler(http_request.method(), http_request.path())?;
+                        stream.write_all(handler.clone().get_response().as_bytes())?;
```

Comment:
We should try to avoid this `clone` call. At the moment we can just change `fn get_response(self)` to `fn get_response(&self)`.

Reply:
It make sense